packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -167,20 +60,325 @@ export const getOutputAmounts = async (params: GetOutputAmountsParams) => {
+        { signal, baseUrl: middlewareUrl },
+    );
+
+    // create zero credentials
+    const zeroAmountCredentialsOut = await middleware.getCredentials(
+        round.amountCredentialIssuerParameters,
+        issuanceData.zeroAmountCredentials,
+        zeroAmountCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+    const zeroVsizeCredentialsOut = await middleware.getCredentials(
+        round.vsizeCredentialIssuerParameters,
+        issuanceData.zeroVsizeCredentials,
+        zeroVsizeCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+
+    // return pairs of new credentials
+    // requested Credentials
+    const output = {
+        amountCredentials: [amountCredentialsOut[0], zeroAmountCredentialsOut[0]],
+        vsizeCredentials: [vsizeCredentialsOut[0], zeroVsizeCredentialsOut[0]],
+    };
+
+    // or change output should be returned to pool
+    // change Credentials
+    const change = {
+        amountCredentials: [amountCredentialsOut[1], zeroAmountCredentialsOut[1]],
+        vsizeCredentials: [vsizeCredentialsOut[1], zeroVsizeCredentialsOut[1]],
+    };
+
+    log(`Output amount credentials: ${output.amountCredentials.map(c => c.value).join(',')}`);
+    log(`Output vsize credentials: ${output.vsizeCredentials.map(c => c.value).join(',')}`);
+    log(`Change amount credentials: ${change.amountCredentials.map(c => c.value).join(',')}`);
+    log(`Change vsize credentials: ${change.vsizeCredentials.map(c => c.value).join(',')}`);
+
+    return {
+        output,
+        change,
+    };
+};
+
+const findCredentialsForTarget = (
+    target: number,
+    credentials: middleware.Credentials[],
+    maxValue: number,
+): [middleware.Credentials, middleware.Credentials] | undefined => {
+    // sort descending. higher possibility for match
+    const sorted = credentials.sort((a, b) => (a.value > b.value ? -1 : 1));
+
+    // find one Credential big enough to cover requested target
+    const bigCredential = sorted.find(cre => cre.value >= target);
+    if (bigCredential) {
+        // try find a pair to produce change Credential used in next iteration
+        // always try to pair with greatest Credential or at least 0 value Credential
+        const changeValue = bigCredential.value - target;
+        const pair = sorted.find(
+            cre => cre !== bigCredential && cre.value + changeValue <= maxValue,
+        );
+        if (pair) {
+            return [bigCredential, pair];
+        }
+        // this should never happen but just in case
+        throw new Error('Missing pair for credential');
+    }
+
+    // one Credential is not enough. try to find pair to cover requested target
+    const candidate = sorted
+        .map(cre => {
+            // always try to pair with greatest Credential to produce greatest possible change
+            const pair = sorted.find(
```

Comment:
This seems like an unnecessary quadratic algorithm. But, given that `sorted` array should be usually small, it may not be worth it to optimize.

Reply:
i think i dont understand, why do you think its unnecessary? how it should be done properly?

Reply:
you were thinking about this?
https://github.com/trezor/trezor-suite/pull/7206/commits/9f5fa28b3fde8c6c2563b35ee4caf6cb50efefca